### PR TITLE
Smallthings

### PR DESCRIPTION
**Pull request by tweisser**
Addressing #5

### CODE REVIEW

File: src/eigenspacedecomposition.jl [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+function row_echelon_form!(A::AbstractMatrix{T}) where T <: FiniteFields.GF
     c, d = size(A)
     l = Int[]
     pos = 0
     i = 0
     for i = 1:d
-        j = findfirst(x -> !iszero(x), A[pos+1:end, i])
-        (j isa Nothing) && continue
+        j = findfirst(x -> !iszero(x), @view A[pos+1:end, i])
+        j == nothing && continue
```
Comment (kalmarek) [inline src/eigenspacedecomposition.jl:8]:
```suggestion
        j === nothing && continue
```

File: src/eigenspacedecomposition.jl [hunk 2]
```diff
@@ -24,21 +24,11 @@ function row_echelon_form!(A::Matrix{T}) where T <: FiniteFields.GF
...
-    n, m = size(A)
-    @assert n == m 
-    B = hcat(deepcopy(A), Matrix{T}(I, n, n))
-    row_echelon_form!(B)
-    return B[:, n+1:end]
-end
-=#
-
-function right_nullspace(M::Matrix{T}) where T <: FiniteFields.GF
+function right_nullspace(M::AbstractMatrix{T}) where T <: FiniteFields.GF
```
Comment (kalmarek) [inline src/eigenspacedecomposition.jl:31]: does the function rely on `T <: GF`? (I don't see how it does)

File: src/eigenspacedecomposition.jl [hunk 3]
```diff
@@ -56,13 +46,13 @@ function right_nullspace(M::Matrix{T}) where T <: FiniteFields.GF
...
 end
 
-function left_eigen(M::Matrix{T}) where T <: FiniteFields.GF
+function left_eigen(M::AbstractMatrix{T}) where T <: FiniteFields.GF
     @assert ==(size(M)...)
-    Id = Matrix{eltype(M)}(I, size(M)...)
+    Id = Matrix{T}(I, size(M)...)
     eigen = Dict{T, typeof(M)}()
     cumdim = 0
     for i in T
```
Comment (kalmarek) [inline src/eigenspacedecomposition.jl:58]:
```suggestion
    for i in T # brute force; TODO: use factorization of characteristic/minimal polynomials
```

File: src/eigenspacedecomposition.jl [hunk 4]
```diff
@@ -81,17 +71,12 @@ function left_eigen(M::Matrix{T}) where T <: FiniteFields.GF
...
-
-function _find_l(M::Matrix{T}) where T <: FiniteFields.GF
+function _find_l(M::AbstractMatrix)
     # this function should be redundant when defining a better structure for echelonized subspaces
     l = Int[]
     for i = 1:size(M, 2)
-        j = findfirst(isone, M[length(l)+1:end,i])
-        if !(j isa Nothing)
+        j = findfirst(isone, @view M[length(l)+1:end,i])
+        if j != nothing
```
Comment (kalmarek) [inline src/eigenspacedecomposition.jl:79]:
```suggestion
        if j !== nothing
```

File: src/eigenspacedecomposition.jl (hunk 2 above)
Comment (tweisser) [reply src/eigenspacedecomposition.jl:31]: Well, to return the trivial solution...we could replace it by eltype...

Comment (tweisser) [reply src/eigenspacedecomposition.jl:31]: Or do you say we should have `T<:Any`?

Comment (kalmarek) [reply src/eigenspacedecomposition.jl:31]: I'd just go with `where T`

Comment (tweisser) [reply src/eigenspacedecomposition.jl:31]: OK,. That's the same as `T<: Any`.

**Pull request merged**